### PR TITLE
feat: implement needs-review PR bypass for assignment cap

### LINKED ISSUE #1469: [Intermediate]: Allow contributors to exceed the open-assignment cap when all their issues are awaiting PR review
### 🧩 Intermediate Friendly

This issue is a good fit for contributors who are already familiar with the Hiero C++ SDK and feel comfortable navigating the codebase.

Intermediate Issues often involve:
- Exploring existing implementations  
- Understanding how different components work together  
- Making thoughtful changes that follow established patterns  

The goal is to support deeper problem-solving while keeping the task clear, focused, and enjoyable to work on.

> [!IMPORTANT]
> ### 🧭 About Intermediate Issues
>
> Intermediate Issues are a great next step for contributors who enjoy digging into the codebase and reasoning about how things work.
>
> These issues often:
> - Involve multiple related files or modules  
> - Encourage investigation and understanding of existing behavior  
> - Leave room for thoughtful implementation choices  
> - Stay focused on a clearly defined goal  
>
> Other kinds of contributions — from beginner-friendly tasks to large system-level changes — are just as valuable and use different labels.

### 👾 Description of the Task

The `/assign` bot currently caps each contributor at **2 open issue assignments** at a time (see `MAX_OPEN_ASSIGNMENTS` in `.github/scripts/commands/assign-comments.js`). This prevents contributors from hoarding issues that they aren't actively working on.

However, this limit creates a frustrating experience for active contributors: if a contributor has submitted PRs for all of their assigned issues and is now simply waiting on review, they are blocked from picking up new work even though they have nothing left to do on their current assignments.

The fix is a bypass: when a contributor is at or above the cap, the bot should check whether every one of their open (non-blocked) assigned issues has an open PR authored by that contributor carrying the `status: needs review` label. If so, the contributor has done all the work they can and is purely waiting on reviewers — they should be allowed to pick up an additional issue.

Important note on where the label lives: `status: needs review` is applied to the **PR**, not to the issue. The issue itself remains `status: in progress` until the PR is merged. The bot must therefore query linked PRs, not issue labels.

The bypass must be **all-or-nothing**: if even one assigned issue does not have a linked open PR with `status: needs review`, the cap still applies.

Relevant files:
```
.github/scripts/commands/assign.js            # enforceAssignmentLimit() — main change
.github/scripts/commands/assign-comments.js   # formatAssignmentLimitExceededComment() — messaging update
.github/scripts/tests/test-assign-bot.js      # new test cases for the bypass
```

### 💡 Proposed Approach

The `enforceAssignmentLimit` function in `assign.js` already counts open non-blocked assignments via `countAssignedIssues`. When that count reaches `MAX_OPEN_ASSIGNMENTS`, the bypass check begins:

1. **Fetch the actual issue objects** (not just a count) for the user's open non-blocked assignments, so their issue numbers are available. The existing `github.rest.issues.listForRepo` call inside `countAssignedIssues` already retrieves full issue objects — a new thin helper (e.g. `listAssignedIssues`) that returns the objects instead of a count would avoid duplicating the pagination logic.

2. **For each assigned issue number**, use the GitHub search API (`github.rest.search.issuesAndPullRequests`) to find open PRs in the repo that are authored by the requesting contributor, carry the `status: needs review` label, and are linked to that issue. GitHub's search supports `linked:<issue-number>` to find PRs linked to a specific issue.

   Example query per issue:
   ```
   repo:owner/repo is:pr is:open author:USERNAME label:"status: needs review" linked:ISSUE_NUMBER
   ```

3. If **every** assigned issue returns at least one matching PR, allow the bypass. If any issue returns zero results (or the search call fails), block as usual.

Update `formatAssignmentLimitExceededComment` (or `buildAssignmentLimitExceededComment`) to explain the bypass condition, so contributors know how to unlock additional assignments.

### 👩‍💻 Implementation Steps

- [ ] Read through `enforceAssignmentLimit` and `countAssignedIssues` in `.github/scripts/commands/assign.js` to understand the current pagination and filtering logic
- [ ] Add a new helper function `listAssignedIssues(github, owner, repo, username)` (in `assign.js` or `helpers/api.js`) that returns the actual open non-blocked issue objects (rather than a count), reusing the same `listForRepo` pagination pattern from `countAssignedIssues` — issues carrying the `status: blocked` label must be excluded, exactly as `countAssignedIssues` already excludes them
- [ ] Add a new helper function `hasNeedsReviewPR(github, owner, repo, username, issueNumber)` that:
  - Calls `github.rest.search.issuesAndPullRequests` with query `repo:owner/repo is:pr is:open author:USERNAME label:"status: needs review" linked:ISSUE_NUMBER`
  - Returns `true` if `total_count > 0`, `false` if zero results, and `null` on API error
- [ ] In `enforceAssignmentLimit`, after confirming `openAssignmentCount >= MAX_OPEN_ASSIGNMENTS`:
  - Fetch open assigned issue objects via `listAssignedIssues`
  - Call `hasNeedsReviewPR` for each issue
  - If any call returns `null` (API error), post the API error comment and return `false`
  - If all calls return `true`, log the bypass and return `true`
  - Otherwise, fall through to the existing limit-exceeded path
- [ ] Update `formatAssignmentLimitExceededComment` in `assign-comments.js` to mention that the limit can be bypassed once all assigned issues have open PRs with `status: needs review`
- [ ] Add test cases in `.github/scripts/tests/test-assign-bot.js` covering:
  - All assigned issues have a linked needs-review PR → bypass succeeds
  - One assigned issue has a linked needs-review PR, another does not → bypass fails, limit-exceeded comment posted
  - No assigned issues have linked needs-review PRs → bypass fails
  - API error on the PR search for one issue → API error comment posted, no bypass
  - Exactly `MAX_OPEN_ASSIGNMENTS` open assignments, all with needs-review PRs → bypass succeeds
- [ ] Run the existing bot test suite to confirm no regressions:
  ```bash
  cd .github/scripts && npm test
  ```
- [ ] Open a pull request with the changes

### ✔️ Acceptance Criteria

- [ ] A contributor with fewer than `MAX_OPEN_ASSIGNMENTS` open issues is unaffected by this change
- [ ] Issues carrying `status: blocked` continue to be excluded from the open-assignment count and from the bypass check, consistent with existing behavior
- [ ] A contributor at or above `MAX_OPEN_ASSIGNMENTS` is allowed to take on a new issue **if and only if** every open (non-blocked) assigned issue has an open PR authored by that contributor with `status: needs review`
- [ ] A contributor at or above the cap where **any** open (non-blocked) assigned issue lacks a qualifying PR is still blocked by the existing limit
- [ ] An API error during the PR search results in the API error comment (not a silent bypass)
- [ ] The limit-exceeded comment explains the bypass condition to the contributor
- [ ] All existing and new test cases pass

---

### 📋 Step-by-Step Contribution Guide

To help keep contributions consistent and easy to review, we recommend following these steps:

- [ ] Comment `/assign` to request the issue
- [ ] Wait for assignment
- [ ] Fork the repository and create a branch
- [ ] Set up the project using the instructions in `README.md`
- [ ] Make the requested changes
- [ ] Sign each commit using `-s -S`
- [ ] Push your branch and open a pull request

Read [Workflow Guide](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/docs/training/workflow.md) for step-by-step workflow guidance.
Read [README.md](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/README.md) for setup instructions.

❗ Pull requests **cannot be merged** without `S` and `s` signed commits.
See the [Signing Guide](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/docs/training/signing.md).

### 🤔 Additional Information

If you have questions while working on this issue, feel free to ask!

You can reach the community and maintainers here: [Hiero-SDK-C++ Discord](https://discord.com/channels/905194001349627914/1337424839761465364)

Whether you need help finding the right file, understanding existing code, or confirming your approach — we're happy to help.

### PR DESCRIPTION
**Pull request by darshit2308**
**Description**:

Enhance the assignment bot to automatically bypass the standard open-assignment limit for highly active contributors. The bot now allows contributors to take on additional issues if they have already submitted pull requests for all their currently assigned work. This ensures top contributors are not artificially bottlenecked while waiting for maintainer reviews.

* Detect when a contributor requests an assignment but is already at or above the cap.
* Fetch all open, non-blocked issues currently assigned to the contributor.
* Use GitHub's GraphQL API to reliably verify if every single assigned issue has a linked, open PR authored by the contributor that carries the `status: needs review` label.
* Automatically approve the new assignment if all existing issues are awaiting review.
* Update the assignment limit-exceeded comment to include a helpful tip informing contributors about how this bypass mechanism works.
* Prevent "vacuous truth" edge cases by explicitly ensuring the assigned issues array is not empty before granting a bypass.
* Add comprehensive integration test coverage for the bypass logic, API failure handling, and the new comment messaging.

**Related issue(s)**:

Fixes #1469

**Notes for reviewer**:

The core bypass logic is injected directly into `enforceAssignmentLimit` in `.github/scripts/commands/assign.js`. 

During implementation, I discovered that GitHub's REST Search API does not support querying specific issue numbers via the `linked:ISSUE_NUMBER` qualifier (it throws a 422 Validation Error). To ensure this bypass is 100% robust and doesn't silently fail, I implemented `hasNeedsReviewPR` in `helpers/api.js` using the GraphQL API (`closedByPullRequestsReferences` on the Issue object) instead of the REST search API. This perfectly maps the PR linkages internally and avoids brittle regex parsing of PR bodies.

Validation:
- Ran `.github/scripts/tests/test-assign-bot.js`
- Result: 41/41 tests passed (35 original + 6 new bypass scenarios)
- Ran the full test suite across all bot scripts
- Result: All 12 test suites passed with 0 failures

**Checklist**

- [x] Documented (Code comments)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: .github/scripts/helpers/api.js [hunk 1]
```diff
@@ -864,6 +864,153 @@ async function countIssuesByAssignee(
...
+              state
+              author { login }
+              labels(first:50) {
+                nodes { name }
+              }
+            }
+          }
+        }
+      }
+    }`;
```
Comment (rwalworth) [inline .github/scripts/helpers/api.js:986]: **thought:** Worth noting that `closedByPullRequestsReferences` only surfaces PRs linked via "Fixes/Closes/Resolves #X" keywords - PRs linked through GitHub's sidebar development menu or via a plain issue mention won't appear here.

This is fine for our workflow since contributors are required to use "Fixes #X" in their PR descriptions, but if a bypass ever fails unexpectedly for an active contributor, this is the first place to look.

**Review by rwalworth — APPROVED** [review]
Looking good, @darshit2308! The implementation is solid - the bypass logic is clean, the vacuous-truth guard is well-reasoned, and the decision to use GraphQL over the REST search API is clearly the right call. I left one small observation below, but nothing blocking. Running the workflows and will merge once they pass!
